src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.hpp

```diff
@@ -0,0 +1,165 @@
+    arrayView1d< real64 const > m_densities;
+    arrayView1d< real64 const > m_viscosities;
+
+    /**
+     * @brief Constructor for the class doing in-kernel fluid property updates
+     * @param[in] densities constant phase densities
+     * @param[in] viscosities constant phase viscosities
+     * @param[in] componentMolarWeight component molecular weights
+     * @param[in] useMass flag to decide whether we return mass or molar densities
+     * @param[in] phaseFraction phase fractions (+ derivatives) in the cell
+     * @param[in] phaseDensity phase mass/molar densities (+ derivatives) in the cell
+     * @param[in] phaseMassDensity phase mass densities (+ derivatives) in the cell
+     * @param[in] phaseViscosity phase viscosities (+ derivatives) in the cell
+     * @param[in] phaseEnthalpy phase enthalpies (+ derivatives) in the cell
+     * @param[in] phaseInternalEnergy phase internal energies (+ derivatives) in the cell
+     * @param[in] phaseCompFraction phase component fractions (+ derivatives) in the cell
+     * @param[in] totalDensity total density in the cell
+     */
+    KernelWrapper( arrayView1d< real64 const > densities,
+                   arrayView1d< real64 const > viscosities,
+                   arrayView1d< real64 const > componentMolarWeight,
+                   bool const useMass,
+                   MultiFluidBase::PhaseProp::ViewType const phaseFraction,
+                   MultiFluidBase::PhaseProp::ViewType const phaseDensity,
+                   MultiFluidBase::PhaseProp::ViewType const phaseMassDensity,
+                   MultiFluidBase::PhaseProp::ViewType const phaseViscosity,
+                   MultiFluidBase::PhaseProp::ViewType const phaseEnthalpy,
+                   MultiFluidBase::PhaseProp::ViewType const phaseInternalEnergy,
+                   MultiFluidBase::PhaseComp::ViewType const phaseCompFraction,
+                   MultiFluidBase::FluidProp::ViewType const totalDensity )
+      : MultiFluidBase::KernelWrapper( std::move( componentMolarWeight ),
+                                       useMass,
+                                       std::move( phaseFraction ),
+                                       std::move( phaseDensity ),
+                                       std::move( phaseMassDensity ),
+                                       std::move( phaseViscosity ),
+                                       std::move( phaseEnthalpy ),
+                                       std::move( phaseInternalEnergy ),
+                                       std::move( phaseCompFraction ),
+                                       std::move( totalDensity )),
+      m_densities( std::move( densities ) ),
+      m_viscosities( std::move( viscosities ) )
+    {}
+
+  };
+
+  /**
+   * @brief Create a KernelWrapper for this fluid model
+   * @return A KernelWrapper instance for this fluid model
+   */
+  KernelWrapper createKernelWrapper() const;
+
+  GEOS_HOST_DEVICE
+  virtual void update( localIndex const k,
```

Comment:
The current pattern is to have this method only on the `KernelWrapper`. Then we can call `InvariantImmiscibleFluid::createKernelWrapper::update()`.